File: backend/node/controller/db_admin.js
const conn = require('../database/db.js');
const prefijoBucket = process.env.PREFIJO_BUCKET;

//========================================== CRUD ARTISTAS ==========================================
function getIdArtista(nombre) {
    return new Promise((resolve, reject) => {
        conn.query('SELECT id_artista FROM Artistas WHERE nombre = ?', nombre, ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.length > 0) {
                    resolve({ status: true, id_artista: result[0].id_artista });
                } else {
                    resolve({ status: false });
                }
             }
        }));
    });
}

function createArtista(nombre, fecha_nacimiento) {
    return new Promise((resolve, reject) => {
        if (fecha_nacimiento != null && fecha_nacimiento != undefined && fecha_nacimiento != '') {
            conn.query('INSERT INTO Artistas (nombre, fecha_nacimiento) VALUES (?, ?)',
            [nombre, fecha_nacimiento], (err, result) => {
                if (err) {
                    reject(err);
                } else {
                    resolve({ status: true, id_artista: result.insertId })
                }
            });
        } else {
            conn.query('INSERT INTO Artistas (nombre) VALUES (?)',
            [nombre, link_foto], (err, result) => {
                if (err) {
                    reject(err);
                } else {
                    resolve({ status: true, id_artista: result.insertId })
                }
            });
        }
    });
}

function readArtistas() {
    return new Promise((resolve, reject) => {
        conn.query('SELECT * FROM Artistas', (async (err, result) => {
            if (err) {
                reject(err);
            } else {
                let artistas = [];
                for (let artista of result) {
                    try {
                        const string_date = `${artista.fecha_nacimiento.getUTCFullYear()}/${artista.fecha_nacimiento.getUTCMonth()+1}/${artista.fecha_nacimiento.getUTCDate()}`;
                        artistas.push({
                            id: artista.id_artista,
                            nombre: artista.nombre,
                            imagen: `${prefijoBucket}Fotos/artistas/${artista.id_artista}.jpg`,
                            nacimiento: string_date
                        })
                    } catch (err) {
                        console.log(err);
                    }
                }
                resolve({ 'artistas': artistas });
            }
        }));
    });
}

function readCancionesArtista(id_artista) {
    return new Promise((resolve, reject) => {
        conn.query(`SELECT c.id_cancion, c.nombre, c.duracion, a.nombre AS artista FROM Canciones c
                    INNER JOIN Artistas a ON a.id_artista = c.id_artista
                    WHERE c.id_artista = ?`, [id_artista], (async (err, result) => {
            if (err) {
                reject(err);
            } else {
                let canciones = [];
                for (let cancion of result) {
                    canciones.push({
                        id: cancion.id_cancion,
                        nombre: cancion.nombre,
                        imagen: `${prefijoBucket}Fotos/canciones/${cancion.id_cancion}.jpg`,
                        duracion: cancion.duracion,
                        artista: cancion.artista
                    })
                }
                resolve({ 'canciones': canciones });
            }
        }));
    });
}

function getNombreArtista(id){
    return new Promise((resolve, reject) => {
        conn.query('SELECT nombre FROM Artistas WHERE id_artista = ?', id, ((err, result) => {
            if (err) {
                reject(err);
            } else {
                resolve({ ok: true });
            }
        }));
    });
}

function updateArtista(id_artista, nombre, fecha) {
    return new Promise((resolve, reject) => {
        conn.query(`UPDATE Artistas 
                    SET nombre = CASE WHEN LENGTH(?) > 0 THEN ? ELSE nombre END,
                        fecha_nacimiento = CASE WHEN LENGTH(?) > 0 THEN ? ELSE fecha_nacimiento END
                    WHERE id_artista = ?`,
                    [nombre, nombre, fecha, fecha, id_artista], ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.changedRows > 0) {
                    resolve({ ok: true })
                } else {
                    resolve({ ok: false })
                }
            }
        }));
    });
}

function deleteArtista(id) {
    return new Promise((resolve, reject) => {
        conn.query('DELETE FROM Artistas WHERE id_artista = ?', id, ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.affectedRows > 0) {
                    resolve({ ok: true })
                } else {
                    resolve({ ok: false })
                }
            }
        }));
    });
}

//========================================= CRUD CANCIONES ==========================================
function getIdCancion(nombre, id_artista) {
    return new Promise((resolve, reject) => {
        conn.query('SELECT id_cancion FROM Canciones WHERE nombre = ? AND id_artista = ?', [nombre, id_artista], ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.length > 0) {
                    resolve({ status: true, id_cancion: result[0].id_cancion });
                } else {
                    resolve({ status: false });
                }
             }
        }));
    });
}

function getIdArtistaCancion(id_cancion) {
    return new Promise((resolve, reject) => {
        conn.query('SELECT id_artista FROM Canciones WHERE id_cancion = ?', id_cancion, ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.length > 0) {
                    resolve({ status: true, id_artista: result[0].id_artista });
                } else {
                    resolve({ status: false });
                }
             }
        }));
    });
}

function createCancion(nombre, duracion, id_artista) {
    return new Promise((resolve, reject) => {     
        conn.query('INSERT INTO Canciones (nombre, duracion, id_artista) VALUES (?, ?, ?)',
        [nombre, duracion, id_artista], (err, result) => {
            if (err) {
                reject(err);
            } else {
                resolve({ status: true, id_cancion: result.insertId })
            }
        });
    });
}

function readCanciones() {
    return new Promise((resolve, reject) => {
        conn.query(`SELECT c.id_cancion, c.nombre, c.duracion, a.nombre AS artista FROM Canciones c
                    INNER JOIN Artistas a ON a.id_artista = c.id_artista`, (async (err, result) => {
            if (err) {
                reject(err);
            } else {
                let canciones = [];
                for (let cancion of result) {
                    canciones.push({
                        id: cancion.id_cancion,
                        nombre: cancion.nombre,
                        imagen: `${prefijoBucket}Fotos/canciones/${cancion.id_cancion}.jpg`,
                        duracion: cancion.duracion,
                        artista: cancion.artista
                    })
                }
                resolve({ 'canciones': canciones });
            }
        }));
    });
}

function updateCancion(id_cancion, nombre, duracion, id_artista) {
    return new Promise((resolve, reject) => {
        conn.query(`UPDATE Canciones 
                    SET nombre = CASE WHEN LENGTH(?) > 0 THEN ? ELSE nombre END,
                        duracion = CASE WHEN ? > 0 THEN ? ELSE duracion END,
                        id_artista = CASE WHEN ? > 0 THEN ? ELSE id_artista END
                    WHERE id_cancion = ?`,
                    [nombre, nombre, duracion, duracion, id_artista, id_artista, id_cancion], ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.changedRows > 0) {
                    resolve({ ok: true })
                } else {
                    resolve({ ok: false })
                }
            }
        }));
    });
}

function deleteCancion(id) {
    return new Promise((resolve, reject) => {
        conn.query('DELETE FROM Canciones WHERE id_cancion = ?', id, ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.affectedRows > 0) {
                    resolve({ ok: true })
                } else {
                    resolve({ ok: false })
                }
            }
        }));
    });
}

//========================================== CRUD ALBUMES ==========================================
function getIdAlbum(nombre, id_artista) {
    return new Promise((resolve, reject) => {
        conn.query('SELECT id_album FROM Albumes WHERE nombre = ? AND id_artista = ?', [nombre, id_artista], ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.length > 0) {
                    resolve({ status: true, id_album: result[0].id_album });
                } else {
                    resolve({ status: false });
                }
             }
        }));
    });
}

function createAlbum(nombre, descripcion, id_artista) {
    return new Promise((resolve, reject) => {     
        conn.query('INSERT INTO Albumes (nombre, descripcion, id_artista) VALUES (?, ?, ?)',
        [nombre, descripcion, id_artista], (err, result) => {
            if (err) {
                reject(err);
            } else {
                resolve({ status: true, id_album: result.insertId })
            }
        });
    });
}

function readAlbumes() {
    return new Promise((resolve, reject) => {
        conn.query(`SELECT alb.id_album, alb.nombre, alb.descripcion, a.nombre AS artista FROM Albumes alb
                    INNER JOIN Artistas a ON a.id_artista = alb.id_artista`, (async (err, result) => {
            if (err) {
                reject(err);
            } else {
                let albumes = [];
                for (let album of result) {
                    albumes.push({
                        id: album.id_album,
                        nombre: album.nombre,
                        descripcion: album.descripcion,
                        imagen: `${prefijoBucket}Fotos/albumes/${album.id_album}.jpg`,
                        artista: album.artista
                    })
                }
                resolve({ 'albums': albumes });
            }
        }));
    });
}

function updateAlbum(id_album, nombre, descripcion) {
    return new Promise((resolve, reject) => {
        conn.query(`UPDATE Albumes 
                    SET nombre = CASE WHEN LENGTH(?) > 0 THEN ? ELSE nombre END,
                        descripcion = CASE WHEN ? > 0 THEN ? ELSE descripcion END
                    WHERE id_album = ?`,
                    [nombre, nombre, descripcion, descripcion, id_album], ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.changedRows > 0) {
                    resolve({ ok: true })
                } else {
                    resolve({ ok: false })
                }
            }
        }));
    });
}

function readCancionesAlbum(id_album) {
    return new Promise((resolve, reject) => {
        conn.query(`SELECT c.id_cancion, c.nombre, c.duracion, a.nombre AS artista FROM Canciones c
                    INNER JOIN Artistas a ON a.id_artista = c.id_artista
                    WHERE c.id_album = ?`, [id_album], (async (err, result) => {
            if (err) {
                reject(err);
            } else {
                let canciones = [];
                for (let cancion of result) {
                    canciones.push({
                        id: cancion.id_cancion,
                        nombre: cancion.nombre,
                        imagen: `${prefijoBucket}Fotos/canciones/${cancion.id_cancion}.jpg`,
                        duracion: cancion.duracion,
                        artista: cancion.artista
                    })
                }
                resolve({ 'canciones': canciones });
            }
        }));
    });
}

function addCancionAlbum(id_cancion, id_album) {
    return new Promise((resolve, reject) => {
        conn.query('UPDATE Canciones SET id_album = ? WHERE id_cancion = ?', [id_album, id_cancion], ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.changedRows > 0) {
                    resolve({ status: true })
                } else {
                    resolve({ status: false })
                }
            }
        }));
    });
}

function deleteCancionAlbum(id_cancion, id_album) {
    return new Promise((resolve, reject) => {
        //Bastaría con comparar el id_cancion pero por seguridad de que no se elimine una canción de otro album
        conn.query('UPDATE Canciones SET id_album = NULL WHERE id_cancion = ? AND id_album = ?', [id_cancion, id_album], ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.changedRows > 0) {
                    resolve({ status: true })
                } else {
                    resolve({ status: false })
                }
            }
        }));
    });
}

function deleteAlbum(id_album) {
    return new Promise((resolve, reject) => {
        conn.query('DELETE FROM Albumes WHERE id_album = ?', id_album, ((err, result) => {
            if (err) {
                reject(err);
            } else {
                if (result.affectedRows > 0) {
                    resolve({ ok: true })
                } else {
                    resolve({ ok: false })
                }
            }
        }));
    });
}

module.exports = {
    getIdArtista,
    getIdArtistaCancion,
    getNombreArtista,
    createArtista,
    readArtistas,
    updateArtista,
    deleteArtista,

    getIdCancion,
    createCancion,
    readCanciones,
    updateCancion,
    deleteCancion,

    getIdAlbum,
    createAlbum,
    updateAlbum,
    readAlbumes,
    deleteAlbum,
    readCancionesAlbum,
    readCancionesArtista,
    addCancionAlbum,
    deleteCancionAlbum
}

File: backend/flask/functions/db_admin.py
import mysql.connector
import os

# Variables de configuración de la base de datos
db_config = {
    'host': 'semi-p1.ct2boqzs8ewg.us-east-1.rds.amazonaws.com',
    'user': 'admin',
    'password': '123456789',
    'database': 'db_semi1_p1',
    'port': 3306,
}

# Variable global para almacenar la conexión
conn = None

# Función para establecer la conexión a la base de datos
def conectar_a_bd():
    global conn
    if conn is None:
        try:
            conn = mysql.connector.connect(**db_config)
            print("Conexión a la base de datos establecida.")
        except mysql.connector.Error as err:
            print(f"Error al conectar a la base de datos: {err}")
    return conn

# Ejecutar la función para establecer la conexión una vez
conectar_a_bd()

prefijoBucket = "https://multimedia-semi1-g9.s3.amazonaws.com/"

#========================================== CRUD ARTISTAS ==========================================
def getIdArtista(nombre):
    cursor = conn.cursor()
    try:
        cursor.execute('SELECT id_artista FROM Artistas WHERE nombre = %s', (nombre,))
        result = cursor.fetchone()
        if result:
            return {'status': True, 'id_artista': result[0]}
        else:
            return {'status': False}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def createArtista(nombre, fecha_nacimiento=None):
    cursor = conn.cursor()
    try:
        if fecha_nacimiento:
            cursor.execute('INSERT INTO Artistas (nombre, fecha_nacimiento) VALUES (%s, %s)', (nombre, fecha_nacimiento))
        else:
            cursor.execute('INSERT INTO Artistas (nombre) VALUES (%s)', (nombre,))
        conn.commit()
        return {'status': True, 'id_artista': cursor.lastrowid}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

def readArtistas():
    cursor = conn.cursor(dictionary=True)
    try:
        cursor.execute('SELECT * FROM Artistas')
        artistas = []
        for artista in cursor:
            try:
                string_date = f"{artista['fecha_nacimiento'].year}/{artista['fecha_nacimiento'].month}/{artista['fecha_nacimiento'].day}"
                artistas.append({
                    'id': artista['id_artista'],
                    'nombre': artista['nombre'],
                    'imagen': f"{prefijoBucket}Fotos/artistas/{artista['id_artista']}.jpg",
                    'nacimiento': string_date
                })
            except Exception as e:
                print(e)
        return {'artistas': artistas}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def readCancionesArtista(id_artista):
    cursor = conn.cursor(dictionary=True)
    try:
        cursor.execute('SELECT c.id_cancion, c.nombre, c.duracion, a.nombre AS artista FROM Canciones c INNER JOIN Artistas a ON a.id_artista = c.id_artista WHERE c.id_artista = %s', (id_artista,))
        canciones = []
        for cancion in cursor:
            canciones.append({
                'id': cancion['id_cancion'],
                'nombre': cancion['nombre'],
                'imagen': f"{prefijoBucket}Fotos/canciones/{cancion['id_cancion']}.jpg",
                'duracion': cancion['duracion'],
                'artista': cancion['artista']
            })
        return {'canciones': canciones}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def getNombreArtista(id):
    cursor = conn.cursor()
    try:
        cursor.execute('SELECT nombre FROM Artistas WHERE id_artista = %s', (id,))
        result = cursor.fetchone()
        if result:
            return {'ok': True}
        else:
            return {'ok': False}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def updateArtista(id_artista, nombre=None, fecha=None):
    cursor = conn.cursor()
    try:
        params = []
        actualizacion = 'SET '

        if nombre:
            actualizacion += 'nombre = %s, '
            params.extend([nombre, nombre])
        if fecha:
            actualizacion += 'fecha_nacimiento = %s, '
            params.extend([fecha, fecha])

        if actualizacion.endswith(', '):
            actualizacion = actualizacion[:-2]  # Eliminar la última coma y espacio

        params.append(id_artista)
        cursor.execute(f'UPDATE Artistas {actualizacion} WHERE id_artista = %s', params)
        conn.commit()

        if cursor.rowcount > 0:
            return {'ok': True}
        else:
            return {'ok': False}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

def deleteArtista(id):
    cursor = conn.cursor()
    try:
        cursor.execute('DELETE FROM Artistas WHERE id_artista = %s', (id,))
        conn.commit()

        if cursor.rowcount > 0:
            return {'ok': True}
        else:
            return {'ok': False}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

#========================================= CRUD CANCIONES ==========================================
def getIdCancion(nombre, id_artista):
    cursor = conn.cursor()
    try:
        cursor.execute('SELECT id_cancion FROM Canciones WHERE nombre = %s AND id_artista = %s', (nombre, id_artista))
        result = cursor.fetchone()
        if result:
            return {'status': True, 'id_cancion': result[0]}
        else:
            return {'status': False}
    except Exception as e:
        raise e
    finally:
        cursor.close()

# ... (Las funciones CRUD de artistas que se proporcionaron anteriormente)

#========================================= CRUD CANCIONES ==========================================
def getIdArtistaCancion(id_cancion):
    cursor = conn.cursor()
    try:
        cursor.execute('SELECT id_artista FROM Canciones WHERE id_cancion = %s', (id_cancion,))
        result = cursor.fetchone()
        if result:
            return {'status': True, 'id_artista': result[0]}
        else:
            return {'status': False}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def createCancion(nombre, duracion, id_artista):
    cursor = conn.cursor()
    try:
        cursor.execute('INSERT INTO Canciones (nombre, duracion, id_artista) VALUES (%s, %s, %s)', (nombre, duracion, id_artista))
        conn.commit()
        return {'status': True, 'id_cancion': cursor.lastrowid}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

def readCanciones():
    cursor = conn.cursor(dictionary=True)
    try:
        cursor.execute('SELECT c.id_cancion, c.nombre, c.duracion, a.nombre AS artista FROM Canciones c INNER JOIN Artistas a ON a.id_artista = c.id_artista')
        canciones = []
        for cancion in cursor:
            canciones.append({
                'id': cancion['id_cancion'],
                'nombre': cancion['nombre'],
                'imagen': f"{prefijoBucket}Fotos/canciones/{cancion['id_cancion']}.jpg",
                'duracion': cancion['duracion'],
                'artista': cancion['artista']
            })
        return {'canciones': canciones}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def updateCancion(id_cancion, nombre=None, duracion=None, id_artista=None):
    cursor = conn.cursor()
    try:
        params = []
        actualizacion = 'SET '

        if nombre:
            actualizacion += 'nombre = %s, '
            params.extend([nombre, nombre])
        if duracion is not None:
            actualizacion += 'duracion = %s, '
            params.extend([duracion, duracion])
        if id_artista is not None:
            actualizacion += 'id_artista = %s, '
            params.extend([id_artista, id_artista])

        if actualizacion.endswith(', '):
            actualizacion = actualización[:-2]  # Eliminar la última coma y espacio

        params.append(id_cancion)
        cursor.execute(f'UPDATE Canciones {actualizacion} WHERE id_cancion = %s', params)
        conn.commit()

        if cursor.rowcount > 0:
            return {'ok': True}
        else:
            return {'ok': False}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

def deleteCancion(id):
    cursor = conn.cursor()
    try:
        cursor.execute('DELETE FROM Canciones WHERE id_cancion = %s', (id,))
        conn.commit()

        if cursor.rowcount > 0:
            return {'ok': True}
        else:
            return {'ok': False}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

#========================================== CRUD ALBUMES ==========================================
def getIdAlbum(nombre, id_artista):
    cursor = conn.cursor()
    try:
        cursor.execute('SELECT id_album FROM Albumes WHERE nombre = %s AND id_artista = %s', (nombre, id_artista))
        result = cursor.fetchone()
        if result:
            return {'status': True, 'id_album': result[0]}
        else:
            return {'status': False}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def createAlbum(nombre, descripcion, id_artista):
    cursor = conn.cursor()
    try:
        cursor.execute('INSERT INTO Albumes (nombre, descripcion, id_artista) VALUES (%s, %s, %s)', (nombre, descripcion, id_artista))
        conn.commit()
        return {'status': True, 'id_album': cursor.lastrowid}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

def readAlbumes():
    cursor = conn.cursor(dictionary=True)
    try:
        cursor.execute('SELECT alb.id_album, alb.nombre, alb.descripcion, a.nombre AS artista FROM Albumes alb INNER JOIN Artistas a ON a.id_artista = alb.id_artista')
        albumes = []
        for album in cursor:
            albumes.append({
                'id': album['id_album'],
                'nombre': album['nombre'],
                'descripcion': album['descripcion'],
                'imagen': f"{prefijoBucket}Fotos/albumes/{album['id_album']}.jpg",
                'artista': album['artista']
            })
        return {'albums': albumes}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def updateAlbum(id_album, nombre=None, descripcion=None):
    cursor = conn.cursor()
    try:
        params = []
        actualizacion = 'SET '

        if nombre:
            actualizacion += 'nombre = %s, '
            params.extend([nombre, nombre])
        if descripcion:
            actualizacion += 'descripcion = %s, '
            params.extend([descripcion, descripcion])

        if actualizacion.endswith(', '):
            actualizacion = actualizacion[:-2]  # Eliminar la última coma y espacio

        params.append(id_album)
        cursor.execute(f'UPDATE Albumes {actualizacion} WHERE id_album = %s', params)
        conn.commit()

        if cursor.rowcount > 0:
            return {'ok': True}
        else:
            return {'ok': False}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

def readCancionesAlbum(id_album):
    cursor = conn.cursor(dictionary=True)
    try:
        cursor.execute('SELECT c.id_cancion, c.nombre, c.duracion, a.nombre AS artista FROM Canciones c INNER JOIN Artistas a ON a.id_artista = c.id_artista WHERE c.id_album = %s', (id_album,))
        canciones = []
        for cancion in cursor:
            canciones.append({
                'id': cancion['id_cancion'],
                'nombre': cancion['nombre'],
                'imagen': f"{prefijoBucket}Fotos/canciones/{cancion['id_cancion']}.jpg",
                'duracion': cancion['duracion'],
                'artista': cancion['artista']
            })
        return {'canciones': canciones}
    except Exception as e:
        raise e
    finally:
        cursor.close()

def addCancionAlbum(id_cancion, id_album):
    cursor = conn.cursor()
    try:
        cursor.execute('UPDATE Canciones SET id_album = %s WHERE id_cancion = %s', (id_album, id_cancion))
        conn.commit()

        if cursor.rowcount > 0:
            return {'status': True}
        else:
            return {'status': False}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

def deleteCancionAlbum(id_cancion, id_album):
    cursor = conn.cursor()
    try:
        cursor.execute('UPDATE Canciones SET id_album = NULL WHERE id_cancion = %s AND id_album = %s', (id_cancion, id_album))
        conn.commit()

        if cursor.rowcount > 0:
            return {'status': True}
        else:
            return {'status': False}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

def deleteAlbum(id_album):
    cursor = conn.cursor()
    try:
        cursor.execute('DELETE FROM Albumes WHERE id_album = %s', (id_album,))
        conn.commit()

        if cursor.rowcount > 0:
            return {'ok': True}
        else:
            return {'ok': False}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        cursor.close()

